### PR TITLE
Frontend MVP for Source Academy Rook

### PR DESCRIPTION
**Pull request by chownces**
### Description

Frontend MVP for Source Academy Rook

#### Status:
- ~To get feedback on UI~
- Need to set up auth providers (Google, Github)
- ~Frontend mockups not updated yet~
- ~'Paths' logic is yet to be refactored~
- ~Assessment 'Grade' has not been fully removed~
- ~To relook at course 'viewability' for admins (it is currently disabled for all when viewable == false)~
- Check that achievements are working properly

Note: To temporarily disable the create course button for deployment

#### Changelog (Knight --> Rook):
- Added multitenant functionality 
    - UI for running multiple courses concurrently on our new backend
    - Users can now create new courses, or enroll in multiple courses

- Updated frontend routing
    - There are now 3 main possible deployment configurations:
        1. Source Academy @ NUS: uses the backend
            - Playground and Sourcecast are now hidden behind the login and are no longer publicly accessible
            - The new SICPJS feature is also hidden behind the login
        2. Playground-only: does not use the backend
            - No notable changes in routing
            - The new SICPJS feature is publicly accessible
        3. Source Academy with GitHub Classroom: an extension of 1 and 2 
            - Enables the GitHub Classroom functionality in either deployment
            - Publicly accessible in either deployment

- Updates to setting of course configuration
    - Previously, the game and achievements were disabled via environment variables in the frontend
    - Now, they are persisted in the backend by course. Different courses can enable/ disable different features at the same time

- New admin panel for course admins
    - Course settings such as course name, user management and toggling of features can now be done in this new admin panel
    - Hardcoded to allow upserts of up to 1000 users at a time

- Changes to testcase types
    - Not much change to the behaviour of these types on the frontend, the changes were mostly due to backend refactoring
    - Previously: Public, Hidden and Private
    - Now: Public, Opaque and Secret

- 'Grade' metric has been deprecated. Assessments will now solely use XP for grading.

### Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### How to test

To be tested with the new backend (https://github.com/source-academy/backend/pull/775)

### Checklist

- [x] I have tested this code
- [ ] I have updated the documentation

### CODE REVIEW

**Comment by chownces** [comment]
Accidentally merged in the previous PR ==. Reissuing it here

File: src/commons/achievement/utils/InsertFakeAchievements.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ function insertFakeAchievements(
       uuid: idString,
       title: assessmentOverview.title,
       ability:
-        assessmentOverview.category === 'Mission' || assessmentOverview.category === 'Path'
+        assessmentOverview.type === 'Missions' || assessmentOverview.type === 'Paths'
```
Comment (chownces) [inline src/commons/achievement/utils/InsertFakeAchievements.ts:56]: probably need the achievements team to remove this hardcode

File: src/commons/achievement/utils/InsertFakeAchievements.ts [hunk 2]
```diff
@@ -71,7 +71,7 @@ function insertFakeAchievements(
       view: {
         coverImage: `${coverImageUrl}/default.png`,
         description: assessmentOverview.shortSummary,
-        completionText: `Grade: ${assessmentOverview.grade} / ${assessmentOverview.maxGrade}`
+        completionText: `XP: ${assessmentOverview.xp} / ${assessmentOverview.maxXp}`
```
Comment (chownces) [inline src/commons/achievement/utils/InsertFakeAchievements.ts:74]: and relook this

File: src/pages/academy/adminPanel/subcomponents/AddUserPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+  const csvRef = React.useRef<CSVReader>(null);
+
+  const validateCsvInput = (data: any[]) => {
+    const processed: UsernameRoleGroup[] = [...users];
+    let hasInvalidInput = false;
+
+    if (data.length + users.length > 1000) {
+      setInvalidCsvMsg('Please limit each upload to 1000 entries!');
+      hasInvalidInput = true;
+    }
```
Comment (chownces) [inline src/pages/academy/adminPanel/subcomponents/AddUserPanel.tsx:95]: still hardcoded to 1000 per upload for now. will be relooked when the super user rights are implemented

File: src/pages/contributors/subcomponents/ContributorsDetails.tsx [hunk 4]
```diff
@@ -210,7 +210,8 @@ class ContributorsDetails extends React.Component {
             </H5>
             <p>
               Below are all contributors to the{' '}
-              <a href={Links.githubOrg}>GitHub repositories of the Source Academy</a>.
+              <a href={Links.githubOrg}>GitHub repositories of the Source Academy</a>. Feel free to
+              join us!
```
Comment (chownces) [inline src/pages/contributors/subcomponents/ContributorsDetails.tsx:214]: requested by prof

File: src/pages/academy/adminPanel/subcomponents/assessmentConfigPanel/DeleteRowCell.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      <Button icon={IconNames.CROSS} onClick={clickHandler} />
+      <Dialog
+        icon={IconNames.WARNING_SIGN}
+        isOpen={isDialogOpen}
+        onClose={() => setIsDialogOpen(false)}
+        title="Warning"
+      >
+        <div className={Classes.DIALOG_BODY}>
+          <p>
+            Are you sure you want to <b>delete</b> the assesment type <i>{props.data.type}</i>?
```
Comment (angelsl) [inline src/pages/academy/adminPanel/subcomponents/assessmentConfigPanel/DeleteRowCell.tsx:37]: assesSment

File: src/pages/academy/adminPanel/subcomponents/assessmentConfigPanel/NumericCell.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import { NumericInput } from '@blueprintjs/core';
+import React from 'react';
+import { AssessmentConfiguration } from 'src/commons/assessment/AssessmentTypes';
+
+type NumericCellProps = OwnProps;
+
+type OwnProps = {
+  data: AssessmentConfiguration;
+  field: AssessmentConfigNumericField;
```
Comment (angelsl) [inline src/pages/academy/adminPanel/subcomponents/assessmentConfigPanel/NumericCell.tsx:9]: Same comment as earlier

File: src/pages/welcome/Welcome.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+const Welcome: React.FC<WelcomeProps> = props => {
+  return (
+    <div className="fullpage">
+      <Card className="fullpage-content">
+        <div style={{ marginTop: '20px', display: 'flex', justifyContent: 'center' }}>
+          <div>
+            <H2>Welcome to Source Academy @ NUS</H2>
+            <div>
+              You have logged in as <strong>{props.name}</strong>. Source Academy @ NUS does not
```
Comment (angelsl) [inline src/pages/welcome/Welcome.tsx:18]: Maybe we should move this "Source Academy @ NUS" title to a configuration env var

File: src/pages/welcome/WelcomeContainer.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import Welcome, { StateProps } from './Welcome';
+
+const mapStateToProps: MapStateToProps<StateProps, {}, OverallState> = state => ({
+  name: state.session.name
+});
+
+const WelcomeContainer = connect(mapStateToProps, null)(Welcome);
+
+export default WelcomeContainer;
```
Comment (angelsl) [inline src/pages/welcome/WelcomeContainer.ts:12]: Let's use hooks (`useSelector`) instead of connected components please..

**Review by angelsl — CHANGES_REQUESTED** [review]
Looks good mostly, just a few minor issues

**Comment by angelsl** [comment]
I left quite a few comments but I think the main issues are

- avoid accessing the Redux store via the `store` global as far as possible (this should be resolved once the course ID is in the URL, future PR)
- avoid hardcoding NUS and NUS-specific terms in new text (this one we can do in a future PR, it's fine)
- avoid using connected components for new components, use hooks e.g. `useSelector` instead
- also, any typos

File: src/commons/achievement/utils/InsertFakeAchievements.ts (hunk 1 above)
Comment (chownces) [reply src/commons/achievement/utils/InsertFakeAchievements.ts:56]: Will be removing this categorisation as discussed. Probably in #1886

**Review by angelsl — APPROVED** [review]
Thanks for all the work!

**Pull request merged**